### DISCUSSION TITLE
dynamic relationships

### DISCUSSION THREAD

**Discussion by miguelgrinberg** [sqlalchemy/sqlalchemy#7372] [Ideas]
Hi Mike & Federico,

Say I have a simple one-to-many relationship between users and blog posts, which I need to be dynamic. With SQLA < 1.4, I can run the dynamic query with a very simple statement:

    posts = user.posts

With 1.4 and onwards, there is nothing that is nearly as simple. The alternative that replicates the above with 1.4+ is, simply put, ugly. Changing the relationship to `noload` as suggested in the docs, then it seems I can only get the posts of a user like this:

    posts = session.scalars(select(Post).where(with_parent(user, User.posts)))

If instead I keep the relationship set to `dynamic`, then I can do it in a much more concise way:

    posts = session.scalars(user.posts.statement)

But this relies on the old `Query` object, and this whole exercise is about not using that and instead using the new query interfaces. So it seems I'm stuck with the long/ugly method?

Without looking at implementation details, to me it seems like the ideal solution would be to have a new `lazy` option (maybe `lazy='statement'`?) that puts the query statement directly in the relationship property, so that there is no need to use `dynamic` just as a hack to fish out the statement from an old `Query` object. Then I would be able to do this:

    posts = session.scalars(user.posts)        # traditional
    posts = await session.scalars(user.posts)  # async

And these look pretty reasonable to me. Any thoughts? Let me know if the `with_parent` example I showed above can be simplified more.

**Comment by CaselIT** [comment]
Hi Miguel,

We had already discussed that, and we do have https://github.com/sqlalchemy/sqlalchemy/issues/7123 open.
We actually documented the `statement` usage also for asyncio in https://github.com/sqlalchemy/sqlalchemy/issues/7082 so for now the current solution is to pass by query.

Regarding #7123 we where thinking about using `.select()`, not having the attribute directly be a select itself. The thinking is that it may make sense having additional functions there

**Reply by miguelgrinberg** [reply]
Oh, great! I missed that in my search. Looks like you will be doing something that is very close to my proposal. 👍

**Reply by zzzeek** [reply]
the docs are showing this as a "legacy feature" still, which it is, but i will update them now to indicate new things are coming b.c. people think with_parent() is the only way to go (was my original plan).

**Reply by zzzeek** [reply]
i've pushed up 29c5fba9ad89e53180f0bd2a0267

**Reply by miguelgrinberg** [reply]
This update makes the current situation perfectly clear. Thanks, Mike!
